tests/test_nq.py

```diff
@@ -0,0 +1,1485 @@
+
+    # Use a random catalog
+    mean_xi_r = data['mean_xi_r']
+    mean_varxi_r = data['mean_varxi_r']
+    var_xi_r = data['var_xi_r']
+
+    print('mean_xi = ',mean_xi_r)
+    print('mean_varxi = ',mean_varxi_r)
+    print('var_xi = ',var_xi_r)
+    print('ratio = ',var_xi_r / mean_varxi_r)
+
+    cat3 = treecorr.Catalog(x=x3, y=y3)
+    rq5 = treecorr.NQCorrelation(corr_params)
+    rq5.process(cat3, cat2)
+    nq5 = nq1.copy()
+    nq5.calculateXi(rq=rq5)
+    print('weight = ',nq5.weight)
+    print('xi = ',nq5.xi)
+    print('varxi = ',nq5.varxi)
+    print('ratio = ',nq5.varxi / var_xi_r)
+    print('pullsq for xi = ',(nq5.xi-mean_xi_r)**2/var_xi_r)
+    print('max pull for xi = ',np.sqrt(np.max((nq5.xi-mean_xi_r)**2/var_xi_r)))
+    np.testing.assert_array_less((nq5.xi-mean_xi_r)**2, 9*var_xi_r)  # < 3 sigma pull
+    np.testing.assert_allclose(nq5.varxi, mean_varxi_r, rtol=0.1)
+
+    # Repeat with patches
+    cat3p = treecorr.Catalog(x=x3, y=y3, patch_centers=cat2p.patch_centers)
+    rq6 = treecorr.NQCorrelation(corr_params)
+    rq6.process(cat3p, cat2p, low_mem=low_mem)
+    nq6 = nq2.copy()
+    nq6.calculateXi(rq=rq6)
+    cov6 = nq6.estimate_cov('jackknife')
+    np.testing.assert_allclose(np.diagonal(cov6), var_xi_r, rtol=0.6)
+
+    # Use a random catalog without patches.
+    rq7 = treecorr.NQCorrelation(corr_params)
+    rq7.process(cat3, cat2p)
+    nq7 = nq4.copy()
+    nq7.calculateXi(rq=rq7)
+    cov7 = nq7.estimate_cov('jackknife')
+    np.testing.assert_allclose(np.diagonal(cov7), var_xi_r, rtol=0.7)
+
+    nq8 = nq2.copy()
+    nq8.calculateXi(rq=rq7)
+    cov8 = nq8.estimate_cov('jackknife')
+    np.testing.assert_allclose(np.diagonal(cov8), var_xi_r, rtol=0.6)
+
+    # Check some invalid actions
+    # Bad var_method
+    with assert_raises(ValueError):
+        nq2.estimate_cov('invalid')
+    # Not run on patches, but need patches
+    with assert_raises(ValueError):
+        nq1.estimate_cov('jackknife')
+    with assert_raises(ValueError):
+        nq1.estimate_cov('sample')
+    with assert_raises(ValueError):
+        nq1.estimate_cov('marked_bootstrap')
+    with assert_raises(ValueError):
+        nq1.estimate_cov('bootstrap')
+    # rq also needs patches (at least for the g part).
+    with assert_raises(RuntimeError):
+        nq2.calculateXi(rq=nq1)
+
+    cat1a = treecorr.Catalog(x=x1[:100], y=y1[:100], npatch=10)
+    cat2a = treecorr.Catalog(x=x2[:100], y=y2[:100], q1=q1[:100], q2=q2[:100], npatch=10)
+    cat1b = treecorr.Catalog(x=x1[:100], y=y1[:100], npatch=2)
+    cat2b = treecorr.Catalog(x=x2[:100], y=y2[:100], q1=q1[:100], q2=q2[:100], npatch=2)
+    nq6 = treecorr.NQCorrelation(corr_params)
+    nq7 = treecorr.NQCorrelation(corr_params)
+    # All catalogs need to have the same number of patches
+    with assert_raises(RuntimeError):
+        nq6.process(cat1a,cat2b)
+    with assert_raises(RuntimeError):
+        nq7.process(cat1b,cat2a)
+
+@timer
+def test_matrix_r():
```

Comment:
@gbernstein, @giuliagiannini 
This test shows how to use the new NQCorrelation class for spin-4 correlations to handle the 2x2 R matrix in BFD.